rfcs/20200525-elasticity-and-fault-tolerance-in-multiworkermirroredstrategy.md

```diff
@@ -0,0 +1,439 @@
+
+
+
+The workflow shown above goes through major and typical situations during elastic training. This occurs within the strategy run function.
+
+* Normal process where the state is saved, but the function runs correctly. 
+* A membership change process that adds a new node (W2, W3). 
+* A communications failure where NetworkExceptions are raised, which initiates a membership change process to remove failing node. 
+* A node failure which causes a node to fail, and propagate by causing the system a network failure on the next iteration. 
+* A normal termination process by OutOfRangeException
+
+### _Variable Sync Process_
+
+The variable sync process is where the chief broadcasts its variables to the other workers to keep them in sync at the beginning. On a new cluster, each node will create their variables locally, then they will synchronize their variables with the chief. The Variable sync process can also be started when there is a new node being added to the cluster. This new node will need to synchronize its variables with the chief. Currently the chief will broadcast to everyone and only the new nodes will actually modify their variables. 
+
+* Broadcast from chief
+    * Catch NetworkError and start Membership Change Process
+* Assign variables from broadcast
+
+**Datasets**
+
+When a new process joins the run call, the dataset of the new nodes and the old nodes will be out of sync. Each dataset operator internally has an iterator object which can be restored through MakeIteratorFromCheckpoint. The chief will send its serialized iterator to be restored on the new nodes. This will bring the new nodes to the last iterated state of the dataset. 
+
+
+### _Membership Change Process_
+
+The membership change process happens when the cluster needs to be modified. This will take place when new nodes are being added to the cluster, or there is a NetworkException raised. The membership change process will attempt to figure out failed nodes, new nodes, and send out cluster spec updates to each worker. 
+
+* Everyone ping chief
+    * timeout for chief to respond
+    * If chief doesn’t respond ping next chief candidate
+* Run discovery and add any new nodes up to max number of nodes
+* Blacklist nodes that did not ping chief (or next candidate)
+* broadcast new cluster spec
+* if number of nodes doesn’t meet minimum threshold, restart membership process
+
+### _Update Cluster_
+
+The update cluster process happens when a node receives a new cluster spec. This cluster spec needs to modify the worker node to discover or blacklist nodes. Currently TensorFlow doesn’t allow for dynamic cluster sizes. There are several identified things that need to be changed in order to allow for dynamic sizes. 
+
+* Collective algorithms will need to be dynamic
+    * There are a node def attributes that need to become inputs that can be set at runtime
+        * **group_size** and **group_key** will need to be Inputs
```

Comment:
You can just rebuild the graph/function without needing to passing new group sizes or keys.

Reply:
Does rebuilding graph/function costs a lot if group size changes? I could imagine that the ring reduce algorithm might need a different packing scheme for different number of workers.

I’m not sure if that’s part of the 3 minutes initialization mentioned above...

Reply:
I think more concrete numbers will be needed to understand the impact of rebuilding the graph. I've been going off other issues which have stated up to 3mins is normal. I will look into getting accurate measurements on several models.

Reply:
We've optimized graph building time and it should be pretty fast these days.

Reply:
Just trying to gather some data on a variety of models. This isn't indepth. 
| Strategy | Model | setup (sec) | step 1 (sec) | step 2 (sec) | step 3 (sec) | step 4 (sec) |
|----------|:-----:|:------:|:------:|:------:|:------:|:-----:|
| mirrored | mnist| 1 g4dn 4 gpus| 9.31 | 0.25 | 0.02 | 0.01|
| mirrored | resnet ctl| 1 g4dn 4 gpus| 30.53 | 0.13 | 0.13 | --|
| mirrored | resnet keras| 1 g4dn 4 gpus| 39.44 | 1.43 | 0.78 | 0.78|
| mirrored | efficientnet| 1 g4dn 4 gpus| 46.58 | 0.91 | 0.29 | 0.28|
| multi-worker | resnet| 2 g4dn 4 gpus| 36.62 | 2.14 | 0.95 | 0.94|
| multi-worker | efficientnet| 2 g4dn 4 gpus| 43.17 | 0.95 | 0.35 | 0.33|

Ill be looking at getting more models like bert and albert to add to this list.

Reply:
Does step 1 here include gpu initialization, variable creation and dataset creation time? I think you can just measure the time of function tracing to measure cost of rebuilding graphs.

Reply:
This is using the TimeHistory keras callback with log_steps=1. 
Which only records on "on_train_batch_begin"
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/training.py#L1089

So GPU, and dataset should be created by now. 
But this should include variable creation, as that is part of the model creation. 
If we had to rebuild the session, won't we have to recreate variables?

Reply:
Whether to build your variables depends on how you implement you solution I think. But I think it is simpler to rebuild variables. According to how you instructment it, it seems to me that your step 0 contains dataset creation/initialization time as well, which could be quite expensive. Rebuilding the dataset in case of worker pool change is necessary if you want to reshard your dataset.

Here we are talking about the necessity to add `group_size` and `group_key` to collective ops. In my opinion, simply recreating the graph should be good enough. So we can just see how expensive it is to recreate the graph.

Reply:
Anyway, changing the collective ops is acceptable if rebuilding the graph turns to be really expensive. My point was that you can start building a quick prototype on top of the existing MWM. This quick prototype won't require changing collective ops if you rebuild the graphs. This prototype would allow us to know better about the potential problems and necessary changes. This also allows potential users to try it out earlier.

Reply:
I understand, thanks!

Reply:
@yuefengz @byronyi  So i have been working on a prototype, Its very raw and its just there to show possible functionality. 
Currently there are two aspects that are working. 
1. Adding a new worker: To add a new worker, you can partially reinitialize the strategy by calling the enable_collective_ops function in the context. Then modify all functions that are loaded in the context with the new group size. (This is only for autographed functions). All ops created in functions should then have the same collective keys, and since they are updated it will have the correct group size. 
2. Resharding Iterators: This involves creating a new API for an elastic iterator. We need to update the iterator internally with a new dataset. This is accomplished by creating a new Operator, this operator saves the old multi-device iterator, creates a new multi-device iterator with a new dataset, then restores the iterator. The iterator restore doesn't change the dataset, only the iterator members. 

The example working right now is in the context of Keras. 
Took the Mnist examples in tensorflow/models, run the model for a long epoch time, and add a new node mid-epoch. This works. Next step is to validate the batch size in the keras context, and update the batch size so the keras logging updates the ETA. 

Are there any other specifics needed, or should i update the RFC with the new findings?